### PR TITLE
Introduced FlowLogic.getFlowContext which provides the flow version a…

### PR DESCRIPTION
**Pull request by shamsasari**
…nd app name of the other side.

I decided to change the app version to app name, which is the CorDapp jar file name, as it should already contain the app version.

### CODE REVIEW

File: docs/source/versioning.rst [hunk 1]
```diff
@@ -31,14 +31,39 @@ for the network.
...
-using the ``version`` property of the ``InitiatingFlow`` annotation. This assigns an integer version number, similar in
-concept to the platform version, which is used in the session handshake process when a flow communicates with another party
-for the first time. The other party will only accept the session request if it, firstly, has that flow loaded, and secondly,
-for the same version (see also :doc:`flow-state-machine`).
-
-.. note:: Currently we don't support multiple versions of the same flow loaded in the same node. This will be possible
-   once we start loading CorDapps in separate class loaders.
+In addition to the evolution of the platform, flows that run on top of the platform can also evolve. It may be that the
+flow protocol between an initiating flow and it's intiated flow changes from one CorDapp release to the next in such as
+way to be backwards incompatible with existing flows. For example, if and sequence of sends and receives needs to change
```
Comment (mikehearn) [inline docs/source/versioning.rst]: if and -> if a

File: docs/source/versioning.rst [hunk 2]
```diff
@@ -31,14 +31,39 @@ for the network.
...
+            val receivedString = if (flowVersionOfOtherParty == 1) {
+                receive<Int>(otherParty).unwrap { it.toString() }
+            } else {
+                receive<String>(otherParty).unwrap { it }
+            }
+        }
+
+The above shows an example evolution of a flow which in the first version was expecting to receive an Int, but then
+in subsequent versions was relaxed to receive a String. This flow is still able to communicate with parties which are
+running the older flow (or rather older CorDapps containing the older flow).
```
Comment (mikehearn) [inline docs/source/versioning.rst]: Can we have some text about the app name here?

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt [hunk 3]
```diff
@@ -7,36 +7,37 @@ import net.corda.core.internal.castIfPossible
...
     val initiatorSessionId: Long
     override val recipientSessionId: Long get() = initiatorSessionId
 }
 
-data class SessionConfirm(override val initiatorSessionId: Long, val initiatedSessionId: Long) : SessionInitResponse
+interface SessionEnd : ExistingSessionMessage
+
+data class SessionInit(val initiatorSessionId: Long,
+                       val initiatingFlowClass: String,
+                       val flowVerison: Int,
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: typo: flowVersion

File: node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt [hunk 4]
```diff
@@ -7,36 +7,37 @@ import net.corda.core.internal.castIfPossible
...
+                       val appName: String,
+                       val firstPayload: Any?) : SessionMessage
+
+data class SessionConfirm(override val initiatorSessionId: Long,
+                          val initiatedSessionId: Long,
+                          val flowVerison: Int,
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: typo: flowVersion

Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SessionMessage.kt]: Gah, last minute refactorings.

File: core/src/main/kotlin/net/corda/core/node/services/NotaryService.kt [hunk 5]
```diff
@@ -21,11 +22,9 @@ abstract class NotaryService : SingletonSerializeAsToken() {
 
     /**
      * Produces a notary service flow which has the corresponding sends and receives as [NotaryFlow.Client].
-     * The first parameter is the client [Party] making the request and the second is the platform version
-     * of the client's node. Use this version parameter to provide backwards compatibility if the notary flow protocol
-     * changes.
+     * The lambda parameter is the client [Party] making the request.
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/services/NotaryService.kt]: Oops

**Comment by shamsasari** [comment]
I've addressed your comments.

I add some docs to say the CoDapp filename can never change once deployed, to make sure counter-party flows have something stable to reply on. That implies the filename should also be in the MANIFEST file, and the node checks that it matches the filename. WDYT?

**Comment by shamsasari** [comment]
Sorry, accidentally closed the PR.

**Comment by mikehearn** [comment]
Well, perhaps we don't need a rule that says it can "never" be changed, just that if you change it you should be sure that nobody is checking the appName field in the context object.

**Pull request merged**